tensorboard/webapp/metrics/views/main_view/card_grid_test.ts

```diff
@@ -0,0 +1,188 @@
+
+import {CardGridComponent} from './card_grid_component';
+import {CardGridContainer} from './card_grid_container';
+import {CardViewComponent} from '../card_renderer/card_view_component';
+import {CardViewContainer} from '../card_renderer/card_view_container';
+import {CardIdWithMetadata} from '../metrics_view_types';
+import {Store} from '@ngrx/store';
+import {MockStore, provideMockStore} from '@ngrx/store/testing';
+import {appStateFromMetricsState} from '../../testing';
+import {State} from '../../../app_state';
+import {selectors as settingsSelectors} from '../../../settings';
+import * as selectors from '../../../selectors';
+
+const scrollElementHeight = 100;
+
+// Utility funciton to allow use of await.
+function timeout(ms: number) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+@Component({
+  selector: 'testing-component',
+  template: `
+    <div cdkScrollable>
+      <div class="placeholder">placeholder</div>
+      <metrics-card-grid
+        [cardIdsWithMetadata]="cardIdsWithMetadata"
+        [cardObserver]="cardObserver"
+        [groupName]="cardObserver"
+      ></metrics-card-grid>
+      <div class="placeholder">placeholder</div>
+    </div>
+  `,
+  styles: [
+    `
+      div {
+        position: fixed;
+        display: block;
+        height: ${scrollElementHeight}px;
+        overflow-y: scroll;
+      }
+      .placeholder {
+        position: relative;
+        height: 700px;
+        display: block;
+      }
+    `,
+  ],
+})
+class TestableComponent {
+  @Input() groupName: string = 'test group name';
+  @Input() cardIdsWithMetadata: CardIdWithMetadata[] = [
+    {
+      cardId: 'card1',
+      plugin: PluginType.SCALARS,
+      tag: 'tagA',
+      runId: null,
+    },
+    {
+      cardId: 'card2',
+      plugin: PluginType.SCALARS,
+      tag: 'tagA/Images',
+      runId: 'run1',
+      sample: 0,
+    },
+    {
+      cardId: 'card3',
+      plugin: PluginType.SCALARS,
+      tag: 'tagB/meow/cat',
+      runId: 'run1',
+      sample: 0,
+    },
+  ];
+  @Input() cardObserver: CardObserver = new CardObserver();
+}
+
+describe('card grid', () => {
+  let store: MockStore<State>;
+  beforeEach(async () => {
+    await TestBed.configureTestingModule({
+      imports: [NoopAnimationsModule, ScrollingModule],
+      declarations: [
+        CardGridComponent,
+        CardGridContainer,
+        TestableComponent,
+        CardViewComponent,
+        CardViewContainer,
+        CardLazyLoader,
+      ],
+      providers: [
+        provideMockStore({
+          initialState: appStateFromMetricsState(),
+        }),
+      ],
+    }).compileComponents();
+
+    store = TestBed.inject<Store<State>>(Store) as MockStore<State>;
+    store.overrideSelector(selectors.getRunColorMap, {});
+  });
+
+  fit('keeps pagination button position when page size changes', async () => {
+    store.overrideSelector(settingsSelectors.getPageSize, 2);
```

Comment:
If I were you, I would specify the `cardIdsWithMetadata` here. Currently, it has three elements and because the pageSize is 2, we should expect next page to fix with less content. Reading this spec alone, this fact is not clear and I need to bounce around in this file, which is a bit unfortunate. You will notice that some tests are needlessly long and what not but this often is the key consideration.

Reply:
I see that being done in other tests with `store.overrideSelector(selectors.getNonEmptyCardIdsWithMetadata,[...])`.
However, they have elements that use that selector to get the cards. My TestingComponent has to inject the cards into the `metrics-card-grid` directly. 

I injected these after creating the component. But that should help the readability and also allow other tests in the future to have a different number of cards. Let me know what you think.